File: C++/프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(string dartResult) {
    int answer = 0, ri = 0;
    vector<int> resultSet;
    string num_str = "";
    
    int i = 0;
    while(i < dartResult.size()) {
        while(45 <= dartResult[i] && dartResult[i] <= 57) {
            num_str += dartResult[i++];
        }
        resultSet.push_back(stoi(num_str));
        while((dartResult[i] < 45 || 57 < dartResult[i]) && i < dartResult.size()) {
            switch(dartResult[i]) {
                case 'D':
                    resultSet[ri] *= resultSet[ri];
                    break;
                case 'T':
                    resultSet[ri] *= (resultSet[ri] * resultSet[ri]);
                    break;
                case '*':
                    resultSet[ri] *= 2;
                    if(ri != 0) resultSet[ri - 1] *= 2;
                    break;
                case '#':
                    resultSet[ri] = -resultSet[ri];
                    break;
            }
            i++;
        }
        ri++; num_str = "";
    }
    
    for(auto& result : resultSet) {
        answer += result;
    }
    
    return answer;
}

File: Python3/프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    answer = []
    number_str = ""
    
    i = 0
    while i < len(dartResult):
        while dartResult[i].isdigit():
            number_str += dartResult[i]
            i += 1
        
        number_str = int(number_str)
        while i < len(dartResult) and not dartResult[i].isdigit():
            if dartResult[i] == 'S':
                number_str **= 1
            elif dartResult[i] == 'D':
                number_str **= 2
            elif dartResult[i] == 'T':
                number_str **= 3
            elif dartResult[i] == '*':
                number_str *= 2
                if answer:
                    answer[-1] *= 2
            elif dartResult[i] == '#':
                number_str = -number_str
            i += 1
        answer.append(number_str)
        # print(number_str)
        number_str = ""
    
    # print(answer)
    return sum(answer)